fastlane_core/lib/fastlane_core/analytics/analytics_session.rb

```diff
@@ -0,0 +1,197 @@
+    attr_accessor :events
+    attr_accessor :fastfile_id
+    attr_accessor :is_fastfile
+    alias fastfile? is_fastfile
+
+    # make this a method so that we can override it in monkey patches
+    def oauth_app_name
+      return 'fastlane-metrics-test-data-v3'
+    end
+
+    def initialize(analytics_ingester_client: AnalyticsIngesterClient.new)
+      require 'securerandom'
+      @session_id = SecureRandom.uuid
+      @client = analytics_ingester_client
+      @events = []
+      @is_fastfile = false
+    end
+
+    def backfill_p_hashes(p_hash: nil)
+      return if p_hash.nil? || @events.count == 0
+      @events.reverse_each do |event|
+        # event[:actor][:name] is the field in which we store the p_hash
+        # to be sent to analytics ingester.
+        # If they are nil, we want to fill them in until we reach
+        # an event that already has a p_hash.
+        event[:actor][:name].nil? ? event[:actor][:name] = p_hash : break
+      end
+    end
+
+    def action_launched(launch_context: nil)
+      backfill_p_hashes(p_hash: launch_context.p_hash)
+
+      builder = AnalyticsEventBuilder.new(
+        oauth_app_name: oauth_app_name,
+        p_hash: launch_context.p_hash,
+        session_id: session_id,
+        action_name: launch_context.action_name
+      )
+
+      @events << builder.launched_event(
+        primary_target_hash: {
+          name: 'fastlane_version',
+          detail: fastlane_version
+        }
+      )
+
+      @events << builder.launched_event(
+        primary_target_hash: {
+          name: 'install_method',
+          detail: install_method
+        }
+      )
+
+      @events << builder.launched_event(
+        primary_target_hash: {
+          name: 'operating_system',
+          detail: operating_system
+        },
+        secondary_target_hash: {
+          name: 'version',
+          detail: operating_system_version
+        }
+      )
+
+      @events << builder.launched_event(
+        primary_target_hash: {
+          name: 'build_tool_version',
+          detail: launch_context.build_tool_version
+        }
+      )
+
+      @events << builder.launched_event(
+        primary_target_hash: {
+          name: 'ci',
+          detail: ci?.to_s
+        }
+      )
+
+      @events << builder.launched_event(
+        primary_target_hash: {
+          name: 'fastfile',
+          detail: fastfile?.to_s
+        },
+        secondary_target_hash: {
+          name: 'fastfile_id',
+          detail: fastfile_id.to_s
+        }
+      )
+
+      @events << builder.launched_event(
+        primary_target_hash: {
+          name: 'platform',
+          detail: launch_context.platform.to_s
+        }
+      )
+
+      @events << builder.launched_event(
+        primary_target_hash: {
+          name: 'ruby_version',
+          detail: ruby_version
+        }
+      )
+    end
+
+    def is_fastfile=(value)
+      if value
+        # If true, update all of the events to reflect
```

Comment:
you write the best comments. I can't wait to go back to make changes and understand why you built things the way you did. Old code is wise code, but clear comments detailing decisions are 💯